<a href="https://colab.research.google.com/github/adgcal/Anjan_collab_R/blob/main/4_bioinfo_R.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# R Curiosity :Bioinformatics 

## Extracting data from PUBMED 


- PubMed is an online repository of references and abstracts of publications in the fields of medicine and life sciences. 
- PubMed is a free resource that is developed and maintained by the National Center for Biotechnology Information (NCBI)
- Alternatively, PubMed can be programmatically queried via the NCBI Entrez E-utilities interface.


In [2]:
install.packages("devtools")

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependencies ‘Rcpp’, ‘httpuv’, ‘xtable’, ‘sourcetools’, ‘later’, ‘promises’, ‘textshaping’, ‘shiny’, ‘downlit’, ‘ragg’, ‘htmlwidgets’, ‘miniUI’, ‘pkgdown’, ‘profvis’, ‘urlchecker’


Warning message in install.packages("devtools"):
“installation of package ‘textshaping’ had non-zero exit status”
Warning message in install.packages("devtools"):
“installation of package ‘ragg’ had non-zero exit status”
Warning message in install.packages("devtools"):
“installation of package ‘pkgdown’ had non-zero exit status”
Warning message in install.packages("devtools"):
“installation of package ‘devtools’ had non-zero exit status”


In [3]:
library(devtools)

Loading required package: usethis



[link text](https://)library(easyPubMed)
# easyPubmed description
- easyPubMed is an R interface to the Entrez Programming Utilities aimed at allowing an easy and smooth programmatic access to PubMed. 
- The package is suitable for batch downloading large volumes of records (via the batch_pubmed_download() function), and comes with a set of functions to perform basic processing of the PubMed query output.
- easyPubMed can request and handle PubMed records in either XML or TXT format. 

# A simple example 
- easyPubMed
- rentrez 


In [4]:
install_github("dami82/easyPubMed")

── R CMD build ─────────────────────────────────────────────────────────────────
* checking for file ‘/tmp/Rtmp7VNj48/remotesfb12609458/dami82-easyPubMed-ea03f36/DESCRIPTION’ ... OK
* preparing ‘easyPubMed’:
* checking DESCRIPTION meta-information ... OK
* checking for LF line-endings in source and make files and shell scripts
* checking for empty or unneeded directories
* building ‘easyPubMed_2.22.tar.gz’



Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [5]:
install.packages("rentrez")

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependency ‘XML’




In [6]:
library(rentrez)

# USING EASYPUBMED 

In [7]:
library(easyPubMed)
my_query <- 'doxyrubicin'
my_entrez_id <- get_pubmed_ids(my_query)
my_abstracts_txt <- fetch_pubmed_data(my_entrez_id, format = "abstract")
cat(my_abstracts_txt,"\n")

1. ACS Appl Mater Interfaces. 2021 Mar 3;13(8):9316-9328. doi:  10.1021/acsami.0c14730. Epub 2020 Oct 22.  Aptamer-Functionalized Upconverting Nanoformulations for Light-Switching  Cancer-Specific Recognition and In Situ Photodynamic-Chemo Sequential  Theranostics.  Jin X(1)(2), Zeng Q(1)(2), Zheng J(1)(2), Xing D(1)(2), Zhang T(1)(2).  Author information: (1)MOE Key Laboratory of Laser Life Science & Institute of Laser Life Science,  College of Biophotonics, South China Normal University, Guangzhou 510631, P. R.  China. (2)Guangdong Provincial Key Laboratory of Laser Life Science, College of  Biophotonics, South China Normal University, Guangzhou 510631, P. R. China.  Biomarker-activatable theranostic formulations offer the potential for removing  specific tumors with a high diagnostic accuracy and a significant  pharmacological effect. Herein, we developed a novel activatable theranostic  nanoformulation UAS-PD [upconversion nanophosphor  (UCNP)-aptamer/ssDNA-pyropheophorbide-a (PPA)

# Using ENTREZ 

In [8]:

# Define the query
my_query <- "doxorubicin"

# Search for articles related to the query and retrieve their IDs
search_results <- entrez_search(db = "pubmed", term = my_query)
# Number of search results 
length(search_results$ids)
my_entrez_id <- search_results$ids
my_abstracts_txt <- entrez_fetch(db = "pubmed", id = my_entrez_id, rettype = "abstract", retmode = "text")
# Read the abstracts file into a character vector
writeLines(my_abstracts_txt, "my_abstracts.txt")
my_abstracts_formatted <- readLines("my_abstracts.txt")
cat(my_abstracts_formatted, sep = "\n")


[1] 20

1. Biomed Pharmacother. 2023 May 2;163:114800. doi: 10.1016/j.biopha.2023.114800.
 Online ahead of print.

Identification of INFG/STAT1/NOTCH3 as γ-Mangostin's potential targets for 
overcoming doxorubicin resistance and reducing cancer-associated fibroblasts in 
triple-negative breast cancer.

Lawal B(1), Wu AT(2), Chen CH(3), T A G(4), Wu SY(5).

Author information:
(1)UPMC Hillman Cancer Center, University of Pittsburgh, Pittsburgh, PA, USA; 
Department of Pathology, University of Pittsburgh, Pittsburgh, PA, USA.
(2)TMU Research Center of Cancer Translational Medicine, Taipei Medical 
University, Taipei 11031, Taiwan; The Ph.D. Program of Translational Medicine, 
College of Medical Science and Technology, Taipei Medical University, Taipei 
11031, Taiwan; Clinical Research Center, Taipei Medical University Hospital, 
Taipei Medical University, Taipei 11031, Taiwan; Graduate Institute of Medical 
Sciences, National Defense Medical Center, Taipei 114, Taiwan.
(3)Department of Surgery, 

In [9]:
# Define the query
my_query <- "side effects of COVID vaccine from pfizer"

# Search for articles related to the query and retrieve their IDs
search_results <- entrez_search(db = "pubmed", term = my_query)
# Number of search results 
length(search_results$ids)
my_entrez_id <- search_results$ids
my_abstracts_txt <- entrez_fetch(db = "pubmed", id = my_entrez_id, rettype = "abstract", retmode = "text")
writeLines(my_abstracts_txt, "pfizer.txt")
my_abstracts_formatted <- readLines("pfizer.txt")
cat(my_abstracts_formatted, sep = "\n")

[1] 20

1. Nutrients. 2023 Apr 20;15(8):1982. doi: 10.3390/nu15081982.

Effects of Gut Microbiome Modulation on Reducing Adverse Health Outcomes among 
Elderly and Diabetes Patients during the COVID-19 Pandemic: A Randomised, 
Double-Blind, Placebo-Controlled Trial (IMPACT Study).

Wong MCS(1)(2), Zhang L(3)(4), Ching JYL(3)(4), Mak JWY(4), Huang J(1)(2), Wang 
S(3)(4), Mok CKP(1), Wong A(3)(4), Chiu OL(3)(4), Fung YT(3)(4), Cheong 
PK(3)(4), Tun HM(1)(3), Ng SC(3)(4), Chan FKL(3)(4).

Author information:
(1)The Jockey Club School of Public Health and Primary Care, Faculty of 
Medicine, The Chinese University of Hong Kong, Hong Kong SAR, China.
(2)Centre for Health Education and Health Promotion, Faculty of Medicine, The 
Chinese University of Hong Kong, Hong Kong SAR, China.
(3)Microbiota I-Center (MagIC), Hong Kong SAR, China.
(4)Centre for Gut Microbiota Research, Faculty of Medicine, The Chinese 
University of Hong Kong, Hong Kong SAR, China.

Gut microbiota is believed to be a major deter

# From *in vivo* , *in vitro* to *in silico*

Bioinformatics came into existence as a result of cumulative efforts of many scientists. But if we have to give the credit to two individuals for setting up the basic platform for bioinformatics we may name Margartette Dayhoff (who discovered consensus sequence) G.N.Ramachandran (who introduced the widely used phi-psi plot) .In the next stage of development , we may name Needleman and Wunch (who introduced global alignment, and a little modification of their dynamical programming algorithm leads to  local alignment). We must mention Atschachul who introduced blast search. A phenomenal growth of the subject was seen, once whole data for genome sequencing started pouring in literature and crystal structure data expanded like never before (thanks to the Cambridge crystallographic repository).  




# The *seqinr* package

In [11]:
install.packages("seqinr")

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependencies ‘pixmap’, ‘sp’, ‘RcppArmadillo’, ‘ade4’, ‘segmented’




## The seqinr package 
is an R package that provides functions for working with biological sequence data. Some of the main things you can do with the seqinr package include:

Read and write sequence data: The read.fasta(), read.abif(), and read.binseq() functions can be used to read sequence data in various file formats, while the write.fasta(), write.abif(), and write.binseq() functions can be used to write sequence data to files.

Manipulate sequence data: The subseq(), reverse(), complement(), and translate() functions can be used to extract subsequences, reverse or complement sequences, and translate nucleotide sequences to protein sequences.

Align sequences: The pairwiseAlignment() function can be used to perform pairwise sequence alignments, while the clustal(), clustalw(), muscle(), and mafft() functions can be used to perform multiple sequence alignments.

Calculate sequence statistics: The nucleotideFrequency(), gcContent(), and seqstat(), functions can be used to calculate various statistics of nucleotide or amino acid sequences, such as the frequency of individual nucleotides or amino acids, the GC content, and the length of the sequence.

Perform sequence analysis: The blast() function can be used to perform BLAST searches against various databases, while the pepstats() and protparam() functions can be used to calculate various physicochemical properties of protein sequences, such as molecular weight, isoelectric point, and hydrophobicity.##

In [13]:
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [14]:
BiocManager::install("Biostrings")


'getOption("repos")' replaces Bioconductor standard repositories, see
'help("repositories", package = "BiocManager")' for details.
Replacement repositories:
    CRAN: https://cran.rstudio.com

Bioconductor version 3.17 (BiocManager 1.30.20), R 4.3.0 (2023-04-21)

Installing package(s) 'BiocVersion', 'Biostrings'

also installing the dependencies ‘bitops’, ‘zlibbioc’, ‘RCurl’, ‘GenomeInfoDbData’, ‘BiocGenerics’, ‘S4Vectors’, ‘IRanges’, ‘XVector’, ‘GenomeInfoDb’


Old packages: 'bit', 'bslib', 'cachem', 'devtools', 'digest', 'dplyr',
  'fontawesome', 'fs', 'gargle', 'gert', 'highr', 'isoband', 'jsonlite',
  'pkgload', 'processx', 'ps', 'rlang', 'roxygen2', 'sass', 'stringr',
  'testthat', 'tidyverse', 'tinytex', 'vctrs', 'viridisLite', 'vroom', 'waldo',
  'whisker', 'xfun', 'xml2', 'zip', 'boot', 'class', 'foreign', 'KernSmooth',
  'MASS', 'nnet'



# Biostring package 
# Sequence Alignment 
\small
Local and global alignment are two popular methods in bioinformatics used for sequence comparison. Both methods aim to identify regions of similarity between two sequences, but they differ in their approach to aligning the sequences.

- Global alignment is used to align two entire sequences from start to end. It attempts to find the best alignment between the entire length of the two sequences, regardless of gaps or mismatches. This type of alignment is useful for comparing two sequences that are expected to be similar across their entire length.

- On the other hand, local alignment focuses on finding the best alignment between subsequences within the two sequences. It aligns only a part of the sequences that have the highest degree of similarity, even if the rest of the sequences are not similar. This type of alignment is useful for comparing two sequences that share regions of high similarity, but are otherwise divergent.

- Another difference between local and global alignment is the scoring method. In global alignment, the Needleman-Wunsch algorithm and in local alignment, the Smith-Waterman algorithm is used


# More about the global and local alignment
## Global alignment
 When comparing two short RNA sequences that are believed to be homologous across their entire length, such as two small RNA molecules that have a conserved function, global alignment would be a suitable method. For example, one could use global alignment to compare the sequences of two microRNAs from different species.

## Local alignment:
 When comparing two short DNA sequences that share a specific sequence motif, but are otherwise variable, local alignment would be a more appropriate method. For example, one could use local alignment to compare the DNA sequences of two different CRISPR spacers to identify conserved regions that might indicate a common target for cleavage by the CRISPR/Cas system.



# Quantification of similarity
\small
 Biostrings package in R that creates a substitution matrix for nucleotide sequences. 
 
 - A substitution matrix assigns a score to every possible substitution of one nucleotide for another, based on how likely it is for that substitution to occur in an evolutionary context.
- The nucleotideSubstitutionMatrix function takes several arguments that allow you to customize the substitution matrix it generates. Here are the arguments you can use:
- match: The score assigned to a match between two identical nucleotides. This value is usually positive, since identical nucleotides are more likely to be conserved through evolution.
- mismatch: The score assigned to a substitution between two different nucleotides. This value is usually negative, since substitutions between different nucleotides are less likely to be conserved.
- baseOnly: A logical value indicating whether to create a 4x4 substitution matrix that only includes the four standard nucleotide bases (A, C, G, T).



In [27]:
suppressPackageStartupMessages(library(Biostrings))
sigma <- nucleotideSubstitutionMatrix(match = 2, mismatch = -1, baseOnly = TRUE)
s1 <- DNAString("GAATTC")
s2 <- DNAString("GATTA")
align <- pairwiseAlignment(s1, s2, substitutionMatrix = sigma, type = "global")
align

Global PairwiseAlignmentsSingleSubject (1 of 1)
pattern: GAATTC
subject: GA-TTA
score: -7 

# Fetching the globin gene and saving it

In [ ]:
library(Biostrings)

## How to get the accession number for the globin gene?

In [16]:
# Load the rentrez package
library(rentrez)

# Define the search term
my_query <- "human beta-globin gene"

# Search for articles related to the query and retrieve their IDs
search_results <- entrez_search(db = "nucleotide", term = my_query)

# Retrieve the first 10 search results
my_entrez_ids <- search_results$ids[1:10]

# Fetch the summary information for the search results
my_summaries <- entrez_summary(db = "nucleotide", id = my_entrez_ids)

# Extract the accession numbers from the summaries
my_accessions <- sapply(my_summaries, function(x) x$accession)

# Print the accession numbers
print(my_accessions)


       170784839       2125203911       2088263913       1531243743 
   "NM_008452.2" "NM_001313903.2" "NM_001313902.2"    "NM_006164.5" 
       926657653        926657647        926657645        926657644 
"NM_001313904.1" "NM_001313901.1" "NM_001313900.1" "NM_001145413.3" 
       926657643       2086959188 
"NM_001145412.3"    "NR_172914.1" 


## Use one of these acessions 

In [24]:

library(rentrez)

# Set the accession number
accession <- "NM_008452.2"

# Use efetch to retrieve the FASTA sequence
seq <- entrez_fetch(db = "nucleotide", id = accession, rettype = "fasta")

# Print the sequence
cat(seq)



>NM_008452.2 Mus musculus Kruppel-like transcription factor 2 (lung) (Klf2), mRNA
ACACACAGTCCCTGCCATGGCGCTCAGCGAGCCTATCTTGCCGTCCTTTGCCACTTTCGCCAGCCCGTGC
GAGCGCGGCCTCCAGGAGCGCTGGCCGCGAAATGAACCCGAGGCGGGCGGCACGGATGAGGACCTAAACA
ACGTGTTGGACTTCATCCTCTCCATGGGATTGGACGGTCTGGGCGCCGAAAATCCTCCCGAGCCCCCGCC
GCAGCCCCCGCCGCCTGCCTTCTACTACCCGGAGCCGGGTGCTCCGCCGCCCTACAGCATCCCCGCGGCC
AGCCTGGGAACAGAGCTGCTGCGCCCCGACCTGGACCCGCCTCAGGGGCCGGCTCTGCACGGCCGCTTCC
TCCTCGCGCCTCCCGGGCGGCTAGTGAAGGCCGAGCCCCCCGAGGTGGACGGCGGCGGCTACGGCTGCGC
TCCGGGCCTGGCCCACGGACCGCGCGGGCTGAAGCTCGAGGGCGCCCCAGGAGCGACAGGTGCATGCATG
CGGGGTCCCGCCGGCCGCCCCCCGCCGCCCCCGGACACGCCGCCGCTCAGCCCCGACGGCCCCCTGCGCA
TCCCGGCGTCCGGTCCCCGCAACCCGTTCCCGCCGCCCTTCGGTCCCGGCCCCAGCTTCGGCGGTCCCGG
CCCCGCGTTGCACTACGGGCCTCCCGCGCCTGGCGCCTTCGGTCTTTTCGAGGACGCGGCGGCAGCGGCG
GCGGCGCTGGGCTTGGCTCCACCTGCCACGCGCGGTCTCCTCACGCCGCCCTCGTCCCCGCTGGAGCTGC
TGGAGGCCAAGCCCAAACGCGGCCGCCGCTCCTGGCCCCGCAAGCGCGCCGCCACACATACTTGCAGCTA
CACCAACTGCGGCAAGACCTACACCAAGAGCTCGCACCTAAAGGCGCATCTGCGTACACACACAGG

## Dengue

- Dengue is a viral disease transmitted by mosquitoes, mainly the Aedes aegypti mosquito.
- It is caused by the Dengue virus, which is a single-stranded RNA virus belonging to the Flaviviridae family.
- Dengue is prevalent in tropical and subtropical regions of the world, and its symptoms include high fever, severe headache, joint and muscle pain, and a rash. 
- There is no specific treatment for Dengue, and supportive care is the mainstay of management.

## H1N1:
- H1N1 is a subtype of the Influenza A virus that causes seasonal flu in humans.
- It is a segmented RNA virus belonging to the Orthomyxoviridae family, and it can infect humans, pigs, and birds.
- H1N1 caused a pandemic in 2009-2010, also known as the swine flu pandemic.
- Symptoms of H1N1 include fever, cough, sore throat, body aches, and fatigue.
- Vaccines are available to prevent H1N1, and antiviral drugs can be used to treat severe cases.

A function to access sequence from NCBI

In [47]:
getncbiseq <- function(accession)
{
require("seqinr") # this function requires the SeqinR R package
# first find which ACNUC database the accession is stored in:
dbs <- c("genbank","refseq","refseqViruses","bacterial")
numdbs <- length(dbs)
for (i in 1:numdbs)
      {
db <- dbs[i]
choosebank(db)
# check if the sequence is in ACNUC database 'db':
resquery <- try(query(".tmpquery", paste("AC=", accession)), silent = TRUE)

if (!(inherits(resquery, "try-error"))) {
  queryname <- "query2"
  thequery <- paste("AC=", accession, sep="")
  query2 <- query(queryname, thequery)
  # see if a sequence was retrieved:
  seq <- getSequence(query2$req[[1]])
  closebank()
  return(seq)
}
closebank()
}
print(paste("ERROR: accession",accession,"was not found"))
}

# The Dengue virus sequence 


The variable dengueseq is a vector containing the nucleotide sequence. Each element of the vector contains one nucleotide of the sequence. Therefore, to print out a certain subsequence of the sequence, we just need to type the name of the vector dengueseq followed by the square brackets containing the indices for those nucleotides. For example, the following command prints out the first 50 nucleotides of the DEN-1 Dengue virus genome sequence

In [48]:
dengueseq <- getncbiseq("NC_001477")

Loading required package: seqinr


Attaching package: ‘seqinr’


The following object is masked from ‘package:Biostrings’:

    translate




# First 20 members of the  Dengue virus sequence 


In [49]:
dengueseq[1:20]


[1] "a" "g" "t" "t" "g" "t" "t" "a" "g" "t" "c" "t" "a" "c" "g" "t" "g" "g" "a"
[20] "c"

In [51]:
write.fasta(names="/content/sample_data/DEN-1", sequences=dengueseq, file.out="den1.fasta")


In [52]:
dengue <- read.fasta(file = "/content/den1.fasta")


Length of a DNA sequence
Once you have retrieved a DNA sequence, we can obtain some simple statistics to describe that sequence, such as the sequence’s total length in nucleotides. In the above example, we retrieved the DEN-1 Dengue virus genome se- quence, and stored it in the vector variable dengueseq. To subsequently obtain the length of the genome sequence, we would use the length() function, typing:


In [53]:
dengueseq <- dengue[[1]]
length(dengueseq)

[1] 10735

 ## Base composition of a DNA sequence


In [54]:
table(dengueseq)

dengueseq
   a    c    g    t 
3426 2240 2770 2299 

# H1N1 Sequence

In [55]:
h1n1seq<-getncbiseq("HM624086")	


In [57]:
write.fasta(names="/content/sample_data/h1n1", sequences=h1n1seq, file.out="h1n1.fasta")


In [58]:
h1n1<-read.fasta(file = "/content/h1n1.fasta")


In [59]:
table(h1n1)

/content/sample_data/h1n1
  a   c   g   t 
603 314 377 407 

In [60]:
h1n1seq<-h1n1[[1]]

# GC contents 

In [61]:
GC(dengueseq)
GC(h1n1seq)


[1] 0.4666977

[1] 0.4062316

# Counting (alternative of Table) 

1.   List item
2.   List item



In [62]:
count(dengueseq,1)
count(h1n1seq,1)


   a    c    g    t 
3426 2240 2770 2299 


  a   c   g   t 
603 314 377 407 

# Sequence alignment

In [68]:
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
    
BiocManager::install("msa")


'getOption("repos")' replaces Bioconductor standard repositories, see
'help("repositories", package = "BiocManager")' for details.
Replacement repositories:
    CRAN: https://cran.rstudio.com

Bioconductor version 3.17 (BiocManager 1.30.20), R 4.3.0 (2023-04-21)

Installing package(s) 'msa'

Old packages: 'bit', 'bslib', 'cachem', 'devtools', 'digest', 'dplyr',
  'fontawesome', 'fs', 'gargle', 'gert', 'highr', 'isoband', 'jsonlite',
  'pkgload', 'processx', 'ps', 'rlang', 'roxygen2', 'sass', 'stringr',
  'testthat', 'tidyverse', 'tinytex', 'vctrs', 'viridisLite', 'vroom', 'waldo',
  'whisker', 'xfun', 'xml2', 'zip', 'boot', 'class', 'foreign', 'KernSmooth',
  'MASS', 'nnet'



In [69]:
library(msa)

In [70]:
seq1<- read.fasta("den1.fasta")
seq2<- read.fasta("h1n1.fasta")

In [73]:
# Perform Clustal alignment
# Two example DNA sequences

sseq1 <- "ATCGCGATCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGC"
sseq2 <- "ATCGCGATCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCAG"

alignment <- msa(c(sseq1, sseq2), method = "ClustalW", type = "dna")
alignment_matrix <- as.matrix(alignment)

# Write the matrix to a file
write.table(alignment_matrix, file = "my_alignment.txt", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

print(alignment_matrix)


use default substitution matrix
     [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13] [,14]
[1,] "A"  "T"  "C"  "G"  "C"  "G"  "A"  "T"  "C"  "G"   "C"   "G"   "C"   "G"  
[2,] "A"  "T"  "C"  "G"  "C"  "G"  "A"  "T"  "C"  "G"   "C"   "G"   "C"   "G"  
     [,15] [,16] [,17] [,18] [,19] [,20] [,21] [,22] [,23] [,24] [,25] [,26]
[1,] "C"   "G"   "C"   "G"   "C"   "G"   "C"   "G"   "C"   "G"   "C"   "G"  
[2,] "C"   "G"   "C"   "G"   "C"   "G"   "C"   "G"   "C"   "G"   "C"   "G"  
     [,27] [,28] [,29] [,30] [,31] [,32] [,33] [,34] [,35] [,36] [,37] [,38]
[1,] "C"   "G"   "C"   "G"   "C"   "G"   "C"   "G"   "C"   "G"   "C"   "G"  
[2,] "C"   "G"   "C"   "G"   "C"   "G"   "C"   "G"   "C"   "G"   "C"   "G"  
     [,39] [,40] [,41] [,42] [,43] [,44] [,45] [,46] [,47] [,48] [,49] [,50]
[1,] "C"   "G"   "C"   "G"   "C"   "G"   "C"   "G"   "C"   "G"   "C"   "G"  
[2,] "C"   "G"   "C"   "G"   "C"   "G"   "C"   "G"   "C"   "G"   "C"   "G"  
     [,51] [,52] [,53] [,54] [,55] 

## RNA-seq analysis in R


<https://combine-australia.github.io/RNAseq-R/00-r-rstudio-intro.html>



Mouse mammary data (counts): https://figshare.com/s/1d788fd384d33e913a2a </p>
Drosophila data (counts): https://figshare.com/s/e08e71c42f118dbe8be6

In [ ]:
list.files()

[1] "den1.fasta"        "h1n1.fasta"        "mouse_c2_v5.rdata"
[4] "sample_data"

Smarter installationof Bioconductor package 
<https://bioconductor.org/packages/release/bioc/>



** Example 1: Biobase installation**

In [ ]:
#library(Biobase)

In [ ]:
#BiocManager::install(c("GenomicFeatures", "AnnotationDbi"))

In [ ]:
$ library(AnnotationDbi)

# Problem 
How many of each of the four nucleotides A, C, T and G, and any other symbols, are there in the Mycobacterium leprae Note: other symbols apart from the four nucleotides A/C/T/G may appear in a sequence. They correspond
to positions in the sequence that are are not clearly one base or another and they are due, for example,
to sequencing uncertainties. or example, the symbol ‘N’ means ‘aNy base’, while ‘R’ means ‘A or G’
(puRine). There is a table of symbols at www.bioinformatics.org/sms/iupac.html.

In [ ]:
acc<- 'NC_002677' 


https://www.ncbi.nlm.nih.gov/nuccore/NC_000962.2?report=fasta

URL=<https://www.ncbi.nlm.nih.gov/nuccore/NC_002677.1/?report=genbank&log$=nuclalign&from=1&to=3414141>

# Quiz Bioinfo 
 - What information about the rabies virus sequence (NCBI accession NC_001542) can you obtain from its annotations.
What does it say in the DEFINITION and ORGANISM fields of its NCBI record? Note: rabies virus is the virus responsible for rabies, which is classified by the WHO as a neglected tropical disease.

- How many nucleotide sequences are there from the bacterium Chlamydia trachomatis in the NCBI Sequence Database? Note: the bacterium Chlamydia trachomatis is responsible for causing trachoma, which is classified by the WHO as a neglected tropical disease.
- How many nucleotide sequences are there from the bacterium Chlamydia trachomatis in the RefSeq part of the NCBI Sequence Database?
- How many nucleotide sequences were submitted to NCBI by Matthew Berriman?
- How many nucleotide sequences from nematode worms are there in the RefSeq Database? Note that several parasitic nematode worms cause neglected tropical diseases, including Brugia malayi and Wucheria bancrofti, which cause lymphatic filariasis; Loa loa, which causes subcutaneous filariasis; Onchocerca volvulus, which causes onchocerciasis; and Necator americanus, which causes soil-transmitted helminthiasis.
- How many nucleotide sequences for collagen genes from nematode worms are there in the NCBI Database?
- How many mRNA sequences for collagen genes from nematode worms are there in the NCBI Database?
- How many protein sequences for collagen proteins from nematode worms are there in the NCBI database?
- What is the accession number for the Trypanosoma cruzi genome in NCBI? Do you see genome se- quences for more than one strain of Trypanosoma cruzi? Note that the Trypanosoma cruzi causes Chagas disease, which is classified as a neglected tropical disease by the WHO.
-How many fully sequenced nematode worm species are represented in the NCBI Genome database?

# Answers




a) What information about the rabies virus sequence (NCBI accession NC_001542) can you obtain from its annotations?

From the annotations of the rabies virus sequence (NCBI accession NC_001542), we can obtain various information such as its DEFINITION, which describes the source organism, length, and molecular type of the sequence. We can also find information about the ORGANISM, which is "Rabies virus" in this case. Additionally, annotations can provide information about the features of the sequence, such as genes, coding regions, and other relevant information.

b) How many nucleotide sequences are there from the bacterium Chlamydia trachomatis in the NCBI Sequence Database?

*See code below*.


c) How many nucleotide sequences are there from the bacterium Chlamydia trachomatis in the RefSeq part of the NCBI Sequence Database?

There are 379 nucleotide sequences from the bacterium Chlamydia trachomatis in the RefSeq part of the NCBI Sequence Database.

d) How many nucleotide sequences were submitted to NCBI by Matthew Berriman?

*See code below*.

e) How many nucleotide sequences from nematode worms are there in the RefSeq Database?

*See code below*.


f) How many nucleotide sequences for collagen genes from nematode worms are there in the NCBI Database?

*See code below*.


g) How many mRNA sequences for collagen genes from nematode worms are there in the NCBI Database?

*See code below*.


h) How many protein sequences for collagen proteins from nematode worms are there in the NCBI database?

*See code below*.


i) What is the accession number for the Trypanosoma cruzi genome in NCBI? Do you see genome sequences for more than one strain of Trypanosoma cruzi?

The accession number for the Trypanosoma cruzi genome in NCBI is GCF_000002325.2. Yes, there are genome sequences available for multiple strains of Trypanosoma cruzi in NCBI.

j) How many fully sequenced nematode worm species are represented in the NCBI Genome database?

As of my knowledge cutoff date of September 2021, there were 68 fully sequenced nematode worm species represented in the NCBI Genome database.

In [74]:
# a
# Load the required packages
library(rentrez)
library(Biostrings)

# Set the NCBI accession number for the Rabies virus genome
acc <- "NC_001542"
rabbies<- getncbiseq(acc)
write.fasta(names="/content/sample_data/rabbies", sequences=rabbies, file.out="rabbies.fasta")


- Visit <https://www.ncbi.nlm.nih.gov/nuccore/NC_001542.1>
- It indicates that the gene "GeneID:1489853" is assigned the locus tag "RABVgp1". 
-  CDS   5418..11846

In [75]:
#b
# Load necessary packages
library(rentrez)

# Search NCBI Sequence Database for Chlamydia trachomatis nucleotide sequences
term <- "Chlamydia trachomatis"
search <- entrez_search(db="nuccore", term=term, retmax=0)
search$count
search$QueryTranslation

[1] 106051

[1] "\"Chlamydia trachomatis\"[Organism] OR Chlamydia trachomatis[All Fields]"

In [76]:
# c
library(rentrez)
search <- entrez_search(db="nucleotide", term="Chlamydia trachomatis[Organism] AND refseq[filter]", retmax=0)
count <- as.integer(search$count)
cat("Number of nucleotide sequences from Chlamydia trachomatis in the RefSeq database:", count)


Number of nucleotide sequences from Chlamydia trachomatis in the RefSeq database: 379

In [77]:
# d
library(rentrez)

# Search NCBI Nucleotide Database for sequences submitted by Matthew Berriman
term <- "Berriman M[Author]"
search <- entrez_search(db="nuccore", term=term, retmax=0)
count <- search$count

# Print the result
cat("Number of nucleotide sequences submitted by Matthew Berriman:", count, "\n")


Number of nucleotide sequences submitted by Matthew Berriman: 648267 


In [78]:
#e
library(rentrez)

# Search for nucleotide sequences from nematode worms in RefSeq Database
term <- "nematoda[Organism] AND refseq[filter]"
search <- entrez_search(db="nuccore", term=term, retmax=0)

# Get the count of nucleotide sequences
count <- search$count
print(paste("Number of nucleotide sequences from nematode worms in RefSeq Database:", count))


[1] "Number of nucleotide sequences from nematode worms in RefSeq Database: 205389"


In [79]:
#f
library(rentrez)

# Search NCBI Nucleotide Database for collagen genes from nematode worms
term <- "collagen AND nematode"
search <- entrez_search(db="nucleotide", term=term, retmax=0)
cat("Number of nucleotide sequences for collagen genes from nematode worms in NCBI Database:", search$count)


Number of nucleotide sequences for collagen genes from nematode worms in NCBI Database: 10735

In [80]:
#g
library(rentrez)

# Set search parameters
term <- "nematoda[orgn] AND collagen AND biomol_mrna[prop]"
search <- entrez_search(db="nucleotide", term=term, retmax=0)

# Print the count result
cat("Number of mRNA sequences for collagen genes from nematode worms in NCBI database:", search$count)


Number of mRNA sequences for collagen genes from nematode worms in NCBI database: 7606

In [81]:
#h
library(rentrez)

# Search NCBI Protein Database for collagen proteins from nematode worms
term <- "nematode collagen protein"
search <- entrez_search(db="protein", term=term, retmax=0)

# Print the number of protein sequences
cat("Number of protein sequences for collagen proteins from nematode worms:",
    format(search$count, big.mark=",", scientific=FALSE))



Number of protein sequences for collagen proteins from nematode worms: 10,795

In [82]:
#i 
library(rentrez)

# Search for the Trypanosoma cruzi genome in the nucleotide database
term <- "Trypanosoma cruzi[Organism]"
search <- entrez_search(db = "nucleotide", term = term)

# Print the accession numbers for the results
if (search$count > 0) {
  res <- entrez_fetch(db = "nucleotide", id = search$ids[1], rettype = "acc")
  cat("Accession number for the T. cruzi genome:", res, "\n")
} else {
  cat("No results found.\n")
}
# ANSWER IS WRONG

Accession number for the T. cruzi genome: OP356744.1
 


In [83]:
library(rentrez)

# Search for all T. cruzi sequences in the nucleotide database
term <- "Trypanosoma"
search <- entrez_search(db = "nucleotide", term = term)

# Filter the results to include only genome sequences
ids <- search$ids[search$ds == "nuccore" & grepl("genome", search$description)]
if (length(ids) > 0) {
  # Print the accession numbers for the genome sequences
  res <- entrez_fetch(db = "nucleotide", id = ids, rettype = "acc")
  cat("Accession numbers for T. cruzi genome sequences:\n", res, "\n")
} else {
  cat("No genome sequences found.\n")
}


No genome sequences found.


# Added Note 
RefSeq (Reference Sequence) is a comprehensive, curated, non-redundant, and well-annotated database of reference sequences for a variety of organisms, including bacteria, viruses, fungi, plants, and animals. The RefSeq database provides a stable reference for genome annotation, gene and protein identification, transcriptomic analysis, and other biological investigations. RefSeq entries are derived from multiple sources, including genomic, transcriptomic, and protein sequence data, and are reviewed and annotated by experts to ensure high quality and accuracy. RefSeq records are regularly updated and maintained, and are widely used as a standard reference for a variety of biological research applications. Thus the number of entries (379)<< (106051)